aten/src/ATen/core/type.cpp

```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  }
+
+  sort_and_filter(types_);
+
+  can_hold_none_ = false;
+  has_free_variables_ = false;
+
+  for (const TypePtr type : types_) {
+    if (type->kind() == NoneType::Kind) {
+      can_hold_none_ = true;
+    }
+    if (type->hasFreeVariables()) {
+      has_free_variables_ = true;
+    }
+  }
+
+}
+
+UnionTypePtr UnionType::create(
+    std::vector<TypePtr> types) {
+  auto union_type = new UnionType(std::move(types));
+  if (auto optional_type = union_type->to_optional()) {
+    return OptionalType::create(std::move(*optional_type));
+  }
+  return UnionTypePtr(std::move(union_type));
+}
+
+bool UnionType::operator==(const Type& rhs) const {
+  if (auto union_rhs = rhs.cast<UnionType>()) {
+    return types_ == union_rhs->types_;
+  }
+  if (auto optional_rhs = rhs.cast<OptionalType>()) {
+    auto optional_lhs = this->to_optional();
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
+    }
+    return false;
+  }
+  return false;
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (rhs_->kind() == NoneType::Kind && can_hold_none_) {
+    return true;
+  }
+  else if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors
+    if (this->types() == union_rhs->types()) {
+      return true;
+    }
+    // Final O(n^2) check since the vector comparison will falsely
+    // flag subtypes as not equal
+    return std::all_of(this->types_.begin(), this->types_.end(),
+      [&](TypePtr this_type) -> bool {
+        return this_type->isSubtypeOf(union_rhs);
+    });
+  }
+  else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    return is_optional_of_type(optional_rhs);
+  }
+  else if (Type::isSubtypeOfExt(rhs_, why_not)) {
+    return true;
+  }
+  else {
+    return false;
+  }
+}
+
+std::string UnionType::str() const {
+  std::stringstream ss;
+  ss << "Union[";
+  for (size_t i = 0; i < types().size(); ++i) {
+    if (i > 0) {
+      ss << ", ";
+    }
+    ss << types()[i]->str();
+  }
+  ss << "]";
+  return ss.str();
+}
+
+std::vector<TypePtr> UnionType::union_of(c10::ArrayRef<TypePtr> rhs_types) const {
+  std::vector<TypePtr> types;
+  types.reserve(this->types().size() + rhs_types.size());
+  types = this->types().vec();
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{types.begin(), types.end()};
+  std::copy_if(rhs_types.begin(), rhs_types.end(),
+            std::back_inserter(types),
+            [&](const TypePtr type) {
+              bool res = !dict.count(type);
+              dict.insert(type);
+              return res;
+            });
+  return types;
+}
+
+std::vector<TypePtr> UnionType::union_of(const UnionTypePtr rhs) const {
+  return union_of(rhs->types());
+}
+
+std::vector<TypePtr> UnionType::intersection_of(c10::ArrayRef<TypePtr> rhs_types) const {
```

Comment:
Yeah - if there's a subtype relation between types, then you need to include subtypes of subtype/supertype pairs when taking an intersection. 

Note that you need to make sure that the UnionType you're intersecting against is canonicalized to not have any internal overlap among its types - members shouldn't be subtypes of each other. Otherwise the results of an intersection can become order-dependent.

You can do the canonicalization at construction time and in `union_of` by being sensitive to subtypes in the opposite way - including supertypes of subtype/supertype pairs, and omitting/removing the subtypes.

Reply:
Good point. I've updated the algorithms, so please let me know if they look good when I push the amended version